WooCommerce/Classes/Authentication/AuthenticationManager.swift

```diff
@@ -449,15 +449,42 @@ private extension AuthenticationManager {
                              with credentials: AuthenticatorCredentials,
                              in navigationController: UINavigationController,
                              onDismiss: @escaping () -> Void) {
-        let viewModel = JetpackErrorViewModel(siteURL: siteURL, onJetpackSetupCompletion: { [weak self] in
-            self?.currentSelfHostedSite = nil
-            self?.presentLoginEpilogue(in: navigationController, for: credentials, onDismiss: onDismiss)
+        let viewModel = JetpackErrorViewModel(siteURL: siteURL, onJetpackSetupCompletion: { [weak self] authorizedEmailAddress in
+            if let site = self?.currentSelfHostedSite {
+                self?.currentSelfHostedSite = WordPressComSiteInfo.updateJetpackInfo(for: site)
+            }
+            guard credentials.wpcom != nil else {
+                return WordPressAuthenticator.showLoginForJustWPCom(from: navigationController, jetpackLogin: true, connectedEmail: authorizedEmailAddress)
+            }
+            // tries re-syncing to get an updated store list
+            // then attempts to present epilogue again
+            ServiceLocator.stores.synchronizeEntities { [weak self] in
+                self?.presentLoginEpilogue(in: navigationController, for: credentials, onDismiss: onDismiss)
+            }
         })
         let installJetpackUI = ULErrorViewController(viewModel: viewModel)
         navigationController.show(installJetpackUI, sender: nil)
     }
 }
 
+private extension WordPressComSiteInfo {
+    /// Copies the information from the specified site and update all Jetpack-related properties for a valid Jetpack connection.
+    ///
+    static func updateJetpackInfo(for site: WordPressComSiteInfo) -> WordPressComSiteInfo {
```

Comment:
nit: it might be because I don't have the full context of how it works with Jetpack operations here, if I understand correctly, maybe some name that indicates the timing of this call like `updatingAfterValidJetpackConnection` could be helpful? `update*` name sounds like it's mutating itself.